### ISSUE TITLE
[5.0] Proposal: Making setNotification and connect APIs cleanup actions in subscriptions disposal

### ISSUE THREAD

**Issue by Kacper20** [Polidea/RxBluetoothKit#173]
This task is more of a discussion about the way we should re-imagine our APIs regarding setting notifications and making connections. It provides an overview on the current implementation and proposes some changes to this. If it’ll be accepted by the maintainers there should be a separate issue created with the detailed design. This two cases are considered together, because our goal is to have consistent APIs. *Remember*: This proposal It’s not necessarily the way to go — it’s only starting point for a discussion. The final solution could be this, sticking with the current approach or even mixing two of them. 

# Design
Currently in each case there are two separate methods: one that is changing state from A to B and the second one that is performing the opposite action. We would like to propose having abstract method:
```
func changeState() -> Observable<X>
```
This method would change state from A to B when subscription to resulting observable is made. The opposite action (transition from B to A) would be made on subscription disposal.
Let’s make this a bit more concrete:

## setNotification changes
The base for all of the notifications related methods is `setNotifyValue(_: Bool, for: Characteristic) -> Observable<Characteristic>`. We would like to propose changing it’s interface to: `setNotifications(on: Characteristic) -> Observable<Characteristic>`. This method would in the implementation set the notification and emit the next when the notification set was successful. This stream should never emit a Complete event. 
In Disposable for this method we would call  `setNotifyValue(false, for: CBCharacteristic)` on an underlying `CBPeripheral` instance. This will disable notifications on this characteristic. Additionally more work related to this observable being connectable will be performed in this piece of code.  Details can be found in `Common considerations for both changes` section.

## connect changes
Currently we have a set of two methods to perform connection related logic: `connect` and `cancelConnection`. To make it more reactive we could make it only one method —`connect`. This method would emit next when connection is made and will never emit a complete. Inside Disposed method we would perform disconnection.
Additionally more work related to this observable being connectable will be performed in this piece of code.  Details can be found in `Common considerations for both changes` section.

## Common considerations for both changes
Because of the nature of those Observable and the possibility that they could be subscribed to multiple times at once we should make them two Connectable Observables. They should be refcounted, so that cleanup is performed only after the last subscription is disposed. In order to implement this we should be storing Observables in some storages internal to BluetoothManager and Peripheral. 

# Affection on the user
It’s source breaking change. If there’ll be a decision to implement this it should be done in 5.0 version.

**Comment by paweljaneczek** [comment]
Added in #212 and #215
